lsp-ui-imenu.el

```diff
@@ -112,62 +149,120 @@
         (when (overlayp lsp-ui-imenu-ov)
           (delete-overlay lsp-ui-imenu-ov))))))
 
+(defun lsp-ui-imenu--move-to-line-beginning ()
+  (-when-let* ((padding (get-char-property (point) 'padding))
+               (bars-count (get-char-property (point) 'bars-count)))
+    (goto-char (+ (* bars-count 3) (line-beginning-position) padding))))
+
 (defvar lsp-ui-imenu--origin nil)
 
 (defun lsp-ui-imenu--put-separator nil
   (let ((ov (make-overlay (point) (point))))
     (overlay-put ov 'after-string (propertize "\n" 'face '(:height 0.6)))))
 
-(defun lsp-ui-imenu--put-kind (title padding color-index)
-  (when (eq lsp-ui-imenu-kind-position 'top)
-    (let ((ov (make-overlay (point) (point)))
-          (color (lsp-ui-imenu--get-color color-index)))
-      (overlay-put
-       ov 'after-string
-       (concat (lsp-ui-imenu--pad " " padding color t)
-               "\n"
-               title
-               (propertize "\n" 'face '(:height 1)))))))
+(defun lsp-ui-imenu--put-toplevel-title (title color-index)
+  (if (eq lsp-ui-imenu-kind-position 'top)
+      (let ((ov (make-overlay (point) (point)))
+            (color (lsp-ui-imenu--get-color color-index)))
+        (overlay-put
+         ov 'after-string
+         (concat (propertize "\n" 'face '(:height 0.6))
+                 (propertize title 'face `(:foreground ,color))
+                 "\n"
+                 (propertize "\n" 'face '(:height 0.6)))))
+    ;; Left placement, title is put with the first sub item. Only put a separator here.
+    (lsp-ui-imenu--put-separator)))
+
+(defun lsp-ui-imenu--put-subtitle (title padding bars bars-count color-index is-last)
+  (let ((ov (make-overlay (point) (point)))
+        (title-color (lsp-ui-imenu--get-color (+ color-index bars-count))))
+    (overlay-put
+     ov 'after-string
+     (concat (lsp-ui-imenu--pad " " padding bars bars-count color-index t is-last)
+             (propertize title 'face `(:foreground ,title-color))
+             (propertize "\n" 'face '(:height 1))))))
+
+(defun lsp-ui-imenu--insert-items (title items padding bars bars-count color-index)
+  "Insert ITEMS for TITLE.
+
+PADDING is the length of whitespaces to the left of the first bar.
+
+BARS is a bool vector of length `lsp-ui-imenu--max-bars'. The ith
+value indicates whether the ith bar from the left is visible.
+
+BARS-COUNTS is the number of bars, including the invisible ones.
+
+COLOR-INDEX is the index of the color of the leftmost bar.
+
+Return the updated COLOR-INDEX."
+  (let ((len (length items)))
+    (--each-indexed items
+      (let ((is-last (eq (1+ it-index) len)))
```

Comment:
I think `eq` is more efficient than `=` when comparing two fixnums but the latter seems more common.

Reply:
Done